lmdeploy/vl/engine.py

```diff
@@ -150,46 +154,93 @@ async def _forward_loop(self):
-        if params is None:
-            params = [{}] * len(inputs)
-        assert len(params) == len(inputs), \
-            'different length of inputs and kwargs'
-        return params
-
-    def forward(self, inputs: List[Image], params: List[Dict] = None):
-        """Model forward."""
-        params = self._init_input_params(inputs, params)
+    def forward(self, messages: List[List[Dict]]) -> Dict:
+        # messages in batch
+        assert all(isinstance(message, List) for message in messages)
+
         time_start = time.perf_counter()
-        func_params = inspect.signature(self.model.forward).parameters
-        func_inputs = [inputs, params] if len(func_params) > 1 else [inputs]
-        outputs = self.model.forward(*func_inputs)
+        outputs, n_image = self.model.forward(messages)
         if isinstance(outputs[0], torch.Tensor):
             outputs = [x.cpu() for x in outputs]
         time_end = time.perf_counter()
-        logger.info(f'ImageEncoder forward {len(inputs)} images, '
+        logger.info(f'ImageEncoder forward {n_image} images, '
                     f'cost {time_end - time_start:.3f}s')
         return outputs
 
-    def infer(self, inputs: List[Image], params: List[Dict] = None):
-        """infer."""
-        params = self._init_input_params(inputs, params)
-        results = self.forward(inputs, params)
-        return results
-
-    async def async_infer(self,
-                          inputs: List[Image],
-                          params: List[Dict] = None):
-        """async infer."""
-        params = self._init_input_params(inputs, params)
-        outputs = asyncio.Queue()
-        item = (inputs, params, outputs)
-        if self.vision_config.thread_safe:
-            self._loop.call_soon_threadsafe(self._que.put_nowait, item)
-        else:
-            self.req_que.put_nowait(item)
-        results = await outputs.get()
-        return results
+    def infer(self, messages: List[Dict]) -> Dict:
+        """perform vision encoding to get a dict, in which there are input_ids,
+        embeddings, embedding_ranges and so on. They will be used by turbomind
+        engine. The key in the dict must be the same defined in turbmoind
+        engine's infer API.
+
+        Args:
+            messages (List[Dict]): user's input in GPT4V format
+        """
+        assert isinstance(messages, List)
+        assert all(isinstance(item, Dict) for item in messages)
+
+        return self.forward(messages)
+
+    async def preprocess(self, messages: List[Dict]) -> List[Dict]:
+        """preprocess multimodal data in the messages."""
+        return self.model.preprocess(messages)
+
+    async def async_infer(self, messages: List[Dict]) -> List[Dict]:
+        """get multimodal embedding.
+
+        Args:
+            messages (List[Dict]): a list of message, which is the output
+            of `preprocess()`
+        """
+        return self.model.forward(messages)
```

Comment:
this function is directly called by `VLAsyncEngine._get_prompt_input` but `self.model.forward` is sync function which will block requests to turbomind.

`self.model.forward(messages)` will forward all images (in messages) in single pass, which may cause oom.